### PR TITLE
Include compara ncbi taxonomy functions in python

### PR DESCRIPTION
**Pull request by CristiGuijarro**
## Description

To go all out Python, we need a Python version of the taxonomy selector we have in Perl. This does that.

**Related JIRA tickets:**
- ENSCOMPARASW-4694
- ENSCOMPARASW-5007

## Overview of changes
All novel stuff, not changes as such, but additions to the codebase.

#### Addition 1
- `taxonomy_selector.py` - as a set of generic util tools related to ncbi taxonomy and making use of `ensembl-py` taxonomy API

#### Change 2
- Included new unit tests, test dbs etc

## Testing
Yes, as above, new unit tests in pytest

## Notes
Another PR will follow this using these utils for more specific RR stuff.

### CODE REVIEW

File: src/python/lib/ensembl/compara/utils/taxonomy/taxon_selector.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Collection of taxonomy methods to configure and extract taxonomy related information.
+
+Typical usage examples::
+
+    from ensembl.compara.utils.taxonomy import *
```
Comment (thiagogenez) [inline src/python/lib/ensembl/compara/utils/taxonomy/taxon_selector.py:21]: Do you have another example without `import *`? Like importing a specific function?

File: src/python/lib/ensembl/compara/utils/taxonomy/taxon_selector.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    """Returns a taxon name if it exists
+
+    Args:
+        taxon_name: Scientific ncbi_taxa_name.name of genome in database
+    """
+    try:
+        Taxonomy.fetch_taxon_by_species_name(session, taxon)
+        return taxon
+    except NoResultFound:
+        return None
```
Comment (thiagogenez) [inline src/python/lib/ensembl/compara/utils/taxonomy/taxon_selector.py:84]: I believe (not sure) PEP recommendations for this would
```
except NoResultFound:
pass

return None
```

File: src/python/lib/ensembl/compara/utils/taxonomy/taxon_selector.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+
+def filter_real_taxon(session: Session, taxon: str):
+    """Returns a taxon name if it exists
+
+    Args:
+        taxon_name: Scientific ncbi_taxa_name.name of genome in database
+    """
+    try:
+        Taxonomy.fetch_taxon_by_species_name(session, taxon)
```
[the remark this answers is not in the document]
Comment (thiagogenez) [reply src/python/lib/ensembl/compara/utils/taxonomy/taxon_selector.py:81]: According to the test below, I believe it will

File: src/python/tests/test_taxon_selector.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+            result3 = match_taxon_to_reference(session, "Canis lupus", list2)
+            assert result1 == exp_out1
+            assert result2 == exp_out2
+            assert result3 == exp_out3
+            with expectation:
+                result4 = match_taxon_to_reference(session, "cake", list2)
+                assert result4 == exp_out1
+
+    @pytest.mark.parametrize(
+        "taxon_1, fake_taxon, expectation", [("mammalia", "cake", "mammalia")]
```
Comment (thiagogenez) [inline src/python/tests/test_taxon_selector.py:106]: maybe adding tests that include `None`, `empty` strings to force unexpected outcomes and exception raises.
This comment also applies to other tests in this file.

**Review by thiagogenez — CHANGES_REQUESTED** [review]
Just a few comments if relevant before merging...

**Review by ipilizota — CHANGES_REQUESTED** [review]
No major comments about the new functions 🎉  Once you have 1 working example of a parametrized test function as a template, you'll get them all 😉

File: src/python/lib/ensembl/compara/utils/taxonomy/taxon_selector.py (hunk 2 above)
Comment (CristiGuijarro) [reply src/python/lib/ensembl/compara/utils/taxonomy/taxon_selector.py:84]: I had to change this to be this way because the pylint/mypy complained when I was initially did it this way :shrug:

File: src/python/lib/ensembl/compara/utils/taxonomy/taxon_selector.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    subdir_names = [f.name for f in os.scandir(rootdir) if f.is_dir(follow_symlinks=True)]
+    valid_taxons = [filter_real_taxon(session, name) for name in subdir_names]
+    taxon_list = list(filter(None, valid_taxons)) # type: List[str]
+    return taxon_list
+
+
+def match_taxon_to_reference(
+    session: Session, taxon_name: str, taxon_list: list
+) -> str:
+    """Returns a taxonomic clade name within the ``taxon_list`` and ``taxon_name`` ancestry
```
[the remark this answers is not in the document]
Comment (CristiGuijarro) [reply src/python/lib/ensembl/compara/utils/taxonomy/taxon_selector.py:56]: This is the only instance in the docstring in which the argument names are used outside of the `Args`, hence they are the only ones in docstrings.

File: src/python/lib/ensembl/compara/utils/taxonomy/taxon_selector.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+    Args:
+        taxon_name: Scientific ncbi_taxa_name.name of genome in database
+        taxon_list: List of clade ncbi_taxa_name.names
+    """
+
```
[the remark this answers is not in the document]
Comment (CristiGuijarro) [reply src/python/lib/ensembl/compara/utils/taxonomy/taxon_selector.py:62]: Yeah, obviously a typo.

File: src/python/tests/test_taxon_selector.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            taxon_id: A real taxon_id
+            useless_int: A fake taxon_id
+            exp_out1: Matching scientific taxon name to `taxon_id`
+            exp_out2: Raises NoResultFound()
+        """
+        with self.dbc.session_scope() as session:
+            result1 = fetch_scientific_name(session, taxon_id)
+            assert result1 == exp_out1
+            with exp_out2:
+                assert fetch_scientific_name(session, useless_int) == exp_out1
```
[the remark this answers is not in the document]
Comment (CristiGuijarro) [reply src/python/tests/test_taxon_selector.py:143]: I've actually got it down elsewhere, this is not my first pytest, I will not need to read the tutorial or sit down with you. I was admittedly a little lazy, and for the limited test case and function I went in this direction, which is also valid, although not preferred.

**Review by thiagogenez — APPROVED** [review]
Things look good to me after reading your comment about pylint/mypy

**Review by dthybert — COMMENTED** [review]
Hi Cristi, everything looks fine to me  a part a very minor point. Just described the parameter session in the documentation of the function collect_taxonomys_from_path. After this I will approve the changes.

**Review by dthybert — APPROVED** [review]
